code/espurna/mcp23s17.cpp

```diff
@@ -0,0 +1,222 @@
+/*
+
+MCP23S17 MODULE
+
+Copyright (C) 2020 by Eddi De Pieri <eddi at depieri dot com>
+Copyright (C) 2016-2017 by Xose Pérez <xose dot perez at gmail dot com>
+Copyright (C) 2016 Plamen Kovandjiev <p.kovandiev@kmpelectronics.eu> & Dimitar Antonov <d.antonov@kmpelectronics.eu>
+
+*/
+
+#include "mcp23s17.h"
+
+#if MCP23S17_SUPPORT
+
+#define CS 0x0F
+
+#define READ_CMD  0x41
+#define WRITE_CMD 0x40 
+
+#define IODIR   0x00
+#define IPOL    0x01
+#define GPINTEN 0x02
+#define DEFVAL  0x03
+#define INTCON  0x04
+#define IOCON   0x05
+#define GPPU    0x06
+#define INTF    0x07
+#define INTCAP  0x08
+#define GPIO    0x09
+#define OLAT    0x0A 
+
+/**
+ * @brief Relay pins.
+ */
+const uint8_t RELAY_PINS[DUMMY_RELAY_COUNT] =
+{ REL1PIN, REL2PIN, REL3PIN, REL4PIN };
```

Comment:
Note that relay count is not always available / can change at runtime. Maybe...
- define max value, gpio_none default
- use runtime vector, push_back on setup